src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -232,23 +227,6 @@ public boolean isApplicable(Class<? extends AbstractProject> jobType) {
             return true;
         }
 
-        @Override
-        public Publisher newInstance(@Nullable StaplerRequest req, @Nonnull JSONObject formData) throws FormException {
```

Comment:
by allowing `null` values we can do away with this, which is arguably a hack around how Jenkins does stuff

Reply:
I thought from your previous comment you were advocating for adding this (option 3) and felt that this option (option 2) was not as ideal. I quite like the fact that using this you can't accidentally set up a step that is invalid.
If you set up an invalid step here, you can only find that out by looking through the logs of your build. The build itself doesn't even fail from it, so it's quite hard to detect

Reply:
I don't like `newInstance` for two reasons
1. (subjective) it is really feels like a hack, I know we do it elsewhere but it really feels like a hack, and I'm not sure how much of API Jenkins core consider that to be, if it is subject to breaking without notice in later versions of Jenkins.
2. I dislike that because I missed filling in a field that all data I've entered *after* (however Jenkins defines after, it seems to be "top to bottom" as ordered on the page) are lost. This to me is the biggest reason. 
3. 
I know most people are trained to hit 'apply' rather than 'save' to avoid this problem, but it still leaves a really bad taste in my mouth because we throw away data the user has entered. There is no real good solution here.

I opted for an error printed to the build log, which I must say think should make it quite easy to determine why it didn't post if you expected it.

Reply:
Sounds like there are good reasons for both options (and both are better than my original implementation). Lets take this offline so we can discuss the pros/cons of each approach a bit more